File: ResourceManager/AppSplashScreen.vb
Imports Microsoft.Extensions.DependencyInjection
Imports ResourceManager.Settings
Imports ResourceManager.Views
Imports VBNetCore.Mvvm.Abstractions

Namespace ResourceManager
    Public Class AppSplashScreen
        Implements IDisposable
        Private productName As String = String.Empty
        Private productVersion As String = String.Empty
        Private _SplashScreenWindow As SplashScreenWindow = Nothing
        ''' 
        ''' 
        Private ReadOnly Property UiExecution As IUiExecution
            Get
                Return Application.ServiceProvider.GetRequiredService(Of IUiExecution)()
            End Get
        End Property
        Public Sub New(ByVal settings As AppSettings)
            productName = settings.AppName
            productVersion = settings.AppVersion
        End Sub
        Public Property IsVisible As Boolean
            Get
                Return Me._SplashScreenWindow IsNot Nothing
            End Get
            Set(ByVal value As Boolean)
                If value Then
                    InitializeSplashScreen()
                Else
                    DisposeSplashScreen()
                End If
            End Set
        End Property

        Public Property Message As String
            Get
                Return _SplashScreenWindow.Message
            End Get
            Set(ByVal value As String)
                SetSplashScreenMessage(If(value, String.Empty))
            End Set
        End Property
        Public Property Progress As Double
            Get
                Return Me._SplashScreenWindow.Progress
            End Get
            Set(ByVal value As Double)
                If value < 0 Then value = 0.0
                SetSplashScreenProgress(value)
            End Set
        End Property

        Private Sub SetSplashScreenProgress(ByVal progress As Double)
            UiExecution.Execute(Sub() Me._SplashScreenWindow.Progress = progress)
        End Sub

        Private Sub InitializeSplashScreen()

            UiExecution.Execute(Sub()
                                    Me._SplashScreenWindow = New SplashScreenWindow()
                                    Me._SplashScreenWindow._productName.Text = productName
                                    Me._SplashScreenWindow._productVersion.Text = productVersion
                                    Me._SplashScreenWindow._message.Text = "Loading.."
                                    Me._SplashScreenWindow.Show()
                                End Sub)
        End Sub

        Private Sub DisposeSplashScreen()
            UiExecution.Execute(Sub()
                                    If Me._SplashScreenWindow Is Nothing Then
                                        Return
                                    End If

                                    Me._SplashScreenWindow.Close()
                                    Me._SplashScreenWindow = Nothing
                                End Sub)
        End Sub
        Private Sub SetSplashScreenMessage(ByVal message As String)
            UiExecution.Execute(Sub() Me._SplashScreenWindow.Message = message)
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            DisposeSplashScreen()
        End Sub
    End Class
End Namespace

File: ResourceManager/AppSplashScreen.cs
using Microsoft.Extensions.DependencyInjection;
using NetCore.Mvvm.Abstractions;
using ResourceManager.Settings;
using ResourceManager.Views;
using System;

namespace ResourceManager
{
    public class AppSplashScreen : IDisposable
    {
        private string productName = string.Empty;
        private string productVersion = string.Empty;
        private SplashScreenWindow splashScreenWindow = null!;
        private IUiExecution UiExecution => App.ServiceProvider.GetRequiredService<IUiExecution>();
        public AppSplashScreen(AppSettings settings)
        {
            productName = settings.AppName;
            productVersion = settings.AppVersion;
        }
        public bool IsVisible
        {
            get => splashScreenWindow != null;
            set
            {
                if (value)
                {
                    InitializeSplashScreen();
                }
                else
                {
                    DisposeSplashScreen();
                }
            }
        }

        public string? Message
        {
            get => splashScreenWindow.Message;
            set => SetSplashScreenMessage(value ?? string.Empty);
        }
        public double? Progress
        {
            get => splashScreenWindow.Progress;
            set => SetSplashScreenProgress(value ?? 0);
        }

        private void SetSplashScreenProgress(double progress)
        {
            UiExecution.Execute(() => { splashScreenWindow.Progress = progress; });
        }

        private void InitializeSplashScreen()
        {

            UiExecution.Execute(() => {
                splashScreenWindow = new SplashScreenWindow();
                splashScreenWindow.productName.Text = productName;
                splashScreenWindow.productVersion.Text = productVersion;
                splashScreenWindow.message.Text = "Loading..";
                splashScreenWindow.Show();
            });
        }

        private void DisposeSplashScreen()
        {
            UiExecution.Execute(() => {
                if (splashScreenWindow == null)
                {
                    return;
                }

                splashScreenWindow.Close();
                splashScreenWindow = null!;
            });
        }
        private void SetSplashScreenMessage(string message)
        {
            UiExecution.Execute(() => { splashScreenWindow.Message = message; });
        }

        public void Dispose()
        {
            DisposeSplashScreen();
        }
    }
}
